### PR TITLE
CIRC-728: Declared lost/aged to lost item: renewal (effect on lost item fees) - ACTUAL COST

### PR DESCRIPTION
**Pull request by yusuf-murodov-epam**
### Purpose
Implementation part of Actual Cost feature
### Approach
Added functionality to refund/close LostItemActualCostFee when renew item, write tests for the change 
### Learning
https://issues.folio.org/browse/CIRC-728

### CODE REVIEW

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java [hunk 1]
```diff
@@ -207,42 +209,50 @@ private CompletableFuture<Result<LostItemFeeRefundContext>> fetchAccountsAndActi
...
     filteredList.add(latestAccount);
     ZonedDateTime creationDate = latestAccount.getCreationDate();
-    List<String> feeFineTypeForSearch = List.of(
-      LOST_ITEM_FEE_TYPE.equals(latestAccount.getFeeFineType())
-        ? LOST_ITEM_PROCESSING_FEE_TYPE
-        : LOST_ITEM_FEE_TYPE);
 
-    getLatestAccount(accounts, feeFineTypeForSearch)
+    List<String> associatedFeeFineAccountType = feeFineTypes.stream()
+      .filter(feeFIneType -> !feeFIneType.equals(latestAccount.getFeeFineType()))
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java:248]:
```suggestion
      .filter(feeFineType -> !feeFineType.equals(latestAccount.getFeeFineType()))
```

File: src/test/java/api/support/fixtures/FeeFineAccountFixture.java [hunk 2]
```diff
@@ -116,15 +131,9 @@ public IndividualResource createManualFeeForLoan(IndividualResource loan, double
...
   }
 
-  private JsonObject getLostItemFeeAccount(UUID loanId) {
-    return accountsClient.getMany(exactMatch("loanId", loanId.toString())
-      .and(exactMatch("feeFineType", "Lost item fee")))
-      .getFirst();
-  }
-
-  private JsonObject getLostItemProcessingFeeAccount(UUID loanId) {
+  private JsonObject getAccountByLoanIdAndFeeFineType(UUID loanId, String feeFineType) {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/support/fixtures/FeeFineAccountFixture.java:134]:
```suggestion
  private JsonObject getAccount(UUID loanId, String feeFineType) {
```

File: src/main/java/org/folio/circulation/domain/FeeFine.java [hunk 3]
```diff
@@ -10,12 +10,12 @@
...
   public static final String LOST_ITEM_FEE_TYPE = "Lost item fee";
-  public static final String LOST_ITEM_PROCESSING_FEE_TYPE = "Lost item processing fee";
   public static final String LOST_ITEM_ACTUAL_COST_FEE_TYPE = "Lost item fee (actual cost)";
+  public static final String LOST_ITEM_PROCESSING_FEE_TYPE = "Lost item processing fee";
 
   private static final Set<String> LOST_ITEM_FEE_TYPES =
-    unmodifiableSet(hashSet(LOST_ITEM_FEE_TYPE, LOST_ITEM_PROCESSING_FEE_TYPE,
-      LOST_ITEM_ACTUAL_COST_FEE_TYPE));
+    unmodifiableSet(hashSet(LOST_ITEM_FEE_TYPE, LOST_ITEM_ACTUAL_COST_FEE_TYPE,
+      LOST_ITEM_PROCESSING_FEE_TYPE));
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/domain/FeeFine.java:18]: it happened when I merged master

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java [hunk 4]
```diff
@@ -207,42 +209,50 @@ private CompletableFuture<Result<LostItemFeeRefundContext>> fetchAccountsAndActi
...
     return getLatestAccount(accounts, feeFineTypes)
       .filter(this::isAccountEligibleForRefund)
-      .map(account -> findRefundableAccounts(account, accounts))
+      .map(account -> findRefundableAccounts(account, accounts, feeFineTypes))
       .orElse(Collections.emptyList());
   }
 
-  private List<Account> findRefundableAccounts(Account latestAccount, Collection<Account> accounts) {
+  private List<Account> findRefundableAccounts(Account latestAccount, Collection<Account> accounts,
+    final List<String> feeFineTypes) {
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java]: removed, now it is unnecessary

File: src/main/java/org/folio/circulation/domain/FeeFine.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/circulation/domain/FeeFine.java:18]: This is pull-request TO master, which means these are your changes

[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/domain/FeeFine.java:18]: Yeah, when I was merging, there appears conflict on the line, the I selected "theirs", which looks nice to me)
or do you want me to return it back?

File: src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java [hunk 5]
```diff
@@ -24,4 +40,82 @@ protected void performActionThatRequiresRefund(AgeToLostFixture.AgeToLostResult
     overrideRenewalFixture.overrideRenewalByBarcode(result.getLoan(),
       servicePointsFixture.cd1().getId());
   }
+
+  @Test
+  void canOverrideRenewalAfterAgeToLostAndRefundsWithLostItemActualCostFee() {
+
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java:46]: Redundant blank line

File: src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java [hunk 6]
```diff
@@ -24,4 +40,82 @@ protected void performActionThatRequiresRefund(AgeToLostFixture.AgeToLostResult
...
+    final double itemFeeActualCost = 10.0;
+    final double itemProcessingFee = 5.0;
+    val policy = lostItemFeePoliciesFixture.facultyStandardPolicy()
+      .withName("shouldChargeOverdueFine")
+      .withActualCost(itemFeeActualCost)
+      .withItemAgedToLostAfterOverdue(Period.minutes(1))
+      .withItemAgedToLostAfterOverdue(minutes(1))
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java:53]: duplicate line

File: src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java [hunk 7]
```diff
@@ -24,4 +40,82 @@ protected void performActionThatRequiresRefund(AgeToLostFixture.AgeToLostResult
...
+      .withPatronBilledAfterItemAgedToLost(minutes(5))
+      .chargeProcessingFeeWhenAgedToLost(itemProcessingFee)
+      .chargeOverdueFineWhenReturned()
+      .withNoFeeRefundInterval();
+
+    val result = ageToLostFixture.createLoanAgeToLostAndChargeFees(policy);
+    createLostItemFeeActualCostAccount(result.getLoan(), itemFeeActualCost);
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java:60]: no, it doesn't

File: src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java [hunk 8]
```diff
@@ -24,4 +40,82 @@ protected void performActionThatRequiresRefund(AgeToLostFixture.AgeToLostResult
...
+
+    assertThat(feeFineActionsClient.getAll(), hasSize(1));
+    assertThat(result.getLoan(), hasLostItemFeeActualCost(isOpen(itemFeeActualCost)));
+
+    feeFineAccountFixture.payLostItemActualCostFee(result.getLoanId(), 3.0);
+    performActionThatRequiresRefund(result, ClockUtil.getZonedDateTime().plusMonths(8));
+
+    assertThat(loansClient.get(result.getLoanId()).getJson().getString("action"),
+      is("renewedThroughOverride"));
+    assertThat(loansClient.get(result.getLoanId()), hasLostItemFeeActualCost(isClosedCancelled(cancellationReason, itemFeeActualCost)));
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java]: fixed

File: src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java [hunk 9]
```diff
@@ -24,4 +40,82 @@ protected void performActionThatRequiresRefund(AgeToLostFixture.AgeToLostResult
...
+    assertThat(result.getLoan(), hasLostItemProcessingFee(isOpen(itemProcessingFee)));
+
+
+    feeFineAccountFixture.payLostItemActualCostFee(result.getLoanId(), 3.0);
+    feeFineAccountFixture.payLostItemProcessingFee(result.getLoanId(), 3.0);
+    performActionThatRequiresRefund(result, ClockUtil.getZonedDateTime().plusMonths(8));
+
+    IndividualResource loan = loansClient.get(result.getLoanId());
+    assertThat(loan.getJson().getString("action"),
+      is("renewedThroughOverride"));
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java]: fixed

File: src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java [hunk 10]
```diff
@@ -24,4 +40,82 @@ protected void performActionThatRequiresRefund(AgeToLostFixture.AgeToLostResult
...
+      .withNoFeeRefundInterval();
+
+    val result = ageToLostFixture.createLoanAgeToLostAndChargeFees(policy);
+    createLostItemFeeActualCostAccount(result.getLoan(), itemFeeActualCost);
+
+    assertThat(feeFineActionsClient.getAll(), hasSize(2));
+    assertThat(result.getLoan(), hasLostItemFeeActualCost(isOpen(itemFeeActualCost)));
+    assertThat(result.getLoan(), hasLostItemProcessingFee(isOpen(itemProcessingFee)));
+
+
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/test/java/api/loans/scenarios/OverrideRenewAgedToLostItemTest.java]: fixed

File: src/test/java/api/loans/scenarios/RefundDeclaredLostFeesTestBase.java [hunk 11]
```diff
@@ -351,7 +352,37 @@ protected void useChargeableRefundableLostItemFee(double itemFee, double process
...
 
+  protected void declareItemLostWithActualCost(double itemFee, double processingFee) {
+    useLostItemPolicy(lostItemFeePoliciesFixture.create(lostItemFeePoliciesFixture
+      .facultyStandardPolicy()
+      .withName(String.format("Test lost policy processing fee %s, item fee %s",
+        processingFee, itemFee))
+      .chargeProcessingFeeWhenDeclaredLost(processingFee)
+      .withActualCost(itemFee).refundFeesWithinMinutes(1)).getId());
+
+    declareItemLost();
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/scenarios/RefundDeclaredLostFeesTestBase.java:363]: Same here, why doesn't decrlaring an item lost create lost item fees automatically?

**Review by OleksandrVidinieiev — APPROVED** [review]

**Pull request merged**